### PR TITLE
starting over

### PR DESCRIPTION
**Pull request by maiamcc**
Porting most of @cjc8's logic, but for me to make it pythonic/testable it's easier if i do it in a fresh file instead of changing it in place.

Here, I've got a new auth flow using env variables instead of a credentials file; the tweet listener and high-level handler flow according to #2; and some tests.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1 +1,33 @@
-# TwitterBot
\ No newline at end of file
+# TwitterBot
+
+## Setup
+- Install requirements: `pip install -r requirements.txt`
+- Make sure you have your auth info for your bot account exported as environment variables
```
Comment (mNov) [inline README.md:5]: maybe would be helpful to link to the Twitter bot documentation? and start with that step, "create a Twitter bot account" or something - just to make it more explicit.

File: README.md [hunk 2]
```diff
@@ -1 +1,33 @@
...
+        export TWITTER_ACCESS_TOKEN_SECRET='wwww'
+        ```
+    2. in your `~/.bash_profile` (or `~/.bashrc` or similar), add:
+        ```
+        source ~/.SECRETS
+        ```
+    3. start a new terminal window or in the current one, run `source ~/.bash_profile` (or `~/.bashrc` or wherever you made the change) 
+### About your bot account setup
+- your bot needs "elevated" access
+- your bot need read/write access: user authentication settings > OAuth 1.0a > app permissions, select "read and write". (The required URL values below can just be nonsense.)
```
Comment (mNov) [inline README.md:20]: nit - typo, should be "needs"

File: README.md [hunk 3]
```diff
@@ -1 +1,33 @@
...
+- when you're done testing, you can run `./clean_up.py` to delete all tweets from your bot.
+
+## Usage
+This bot can make database lookups by name or badge number.
+
+If given a badge number, will look for an exact match and return a failure message if can't find one. (Bot will attempt to validate the badge number, and complain if it's impossible—e.g. if bot gets passed a numeric string but it's too long to be a valid badge number.)
+
+If given a name, will search the DB (using substring search, i.e. searching for `WHERE name LIKE '%name%'`). If the name uniquely matches a row, return that row; otherwise print info about all matching names with their badge numbers, and request user to search by unique name or badge.
+
+Any request that isn't obviously a badge number (i.e. isn't purely numeric) will be treated as a name query.
```
Comment (mNov) [inline README.md:33]: very helpful readme!

File: bot.py [hunk 4]
```diff
@@ -0,0 +1,112 @@
+#! /usr/bin/env python
+
+import os
+import re
+
+import requests
+import tweepy
+import urllib
+
+API_URL = os.environ.get('COP_API_URL', 'http://localhost:3000/api/v1/cops')
```
Comment (mNov) [inline bot.py:10]: should there be something about this in the README too (i.e. running the API server)? maybe not, i'm not sure if that's a good idea/necessary

File: bot.py [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+                         access_token, access_token_secret, **kwargs)
+
+    def start(self):
+        """The main loop of the app: stream incoming tweets."""
+        self.filter(track=[f'@{self.user.screen_name}'])
+
+    def on_status(self, status: Status):
+        """Extends tweepy.Stream, will be called on every incoming Tweet"""
+        if status.author.id != self.user.id:  # only respond to tweets from users that are not the bot
+            print(f'Got tweet from @{status.author.screen_name}: {status.text}')
```
Comment (mNov) [inline bot.py:57]: where will this be printed?

File: test_bot.py [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+from test_utils import tweet_with_text
+
+fake_environ = {
+    'TWITTER_API_KEY': 'foo',
+    'TWITTER_API_SECRET_KEY': 'bar',
+    'TWITTER_ACCESS_TOKEN': 'baz',
+    'TWITTER_ACCESS_TOKEN_SECRET': 'quux',
+}
+
+
```
Comment (mNov) [inline test_bot.py:17]: great tests here!

File: bot.py (hunk 5 above)
Comment (maiamcc) [reply bot.py:57]: on whatever server we're running the bot. We can decide before we actually host this somewhere that they're more annoying than helpful, but I want to leave them for now for debugging purposes. In a future PR I can put them in a `logger` object that supports multiple log levels, that'll take like 2 minutes.

Comment (maiamcc) [reply bot.py:57]: (But like, whoever is tweeting at us won't see this, if that's your question

Comment (mNov) [reply bot.py:57]: that makes sense, I was wondering if this was for logging/debugging. sounds good!

File: bot.py (hunk 4 above)
Comment (maiamcc) [reply bot.py:10]: that's... a great question. I will put in a line about running it, but I'm hesitant to explicitly link to the other repo; @schrockblock do you have thoughts about this?

**Comment by maiamcc** [comment]
@mNov think I addressed all your feedback, lmk if there's anything else. If not, do you feel good about approving? (Or did you want to wait for someone else to review?)

File: README.md [hunk 7]
```diff
@@ -1 +1,36 @@
-# TwitterBot
\ No newline at end of file
+# TwitterBot
+
+## Setup
+- Install requirements: `pip install -r requirements.txt`
+- [Register for Twitter developer access](https://developer.twitter.com/en/portal/petition/essential/basic-info) and create a new bot with an api key and secret, and an access token and secret (the first two should be generated for you on bot creation; you may need to click a button to generate the latter two)
```
Comment (mNov) [inline README.md:5]: nice 👍

File: README.md [hunk 8]
```diff
@@ -1 +1,36 @@
...
+    3. start a new terminal window or in the current one, run `source ~/.bash_profile` (or `~/.bashrc` or wherever you made the change)
+- run the backing api on localhost:3000 (this lives in a separate repo)
+
+### About your bot account setup
+- your bot needs "elevated" access
+- your bot needs read/write access: user authentication settings > OAuth 1.0a > app permissions, select "read and write". (The required URL values below can just be nonsense.)
+
+### About testing your bot
+- whatever account is tweeting at your bot may NOT be private/have its tweets protected, or else the API won't be able to pick up its tweets to respond (even if that account and your bot account follow each other)
+- when you're done testing, you can run `./clean_up.py` to delete all tweets from your bot.
```
Comment (mNov) [inline README.md:27]: this is great, btw!

File: clean_up.py [hunk 9]
```diff
@@ -0,0 +1,29 @@
...
+
+from bot import Credentials
+
+creds = Credentials()
+api = creds._auth()
+
+deleted = 0
+failed = 0
+
+for status in tweepy.Cursor(api.user_timeline).items():
```
Comment (mNov) [inline clean_up.py:15]: very cool feature here 👍

**Review by mNov — APPROVED** [review]
looks good to me! still wondering about whether to link the 2 repos, but that can be addressed later

File: README.md (hunk 8 above)
Comment (cjc8) [reply README.md:27]: Really really great!!

**Review by cjc8 — APPROVED** [review]
LGTM!

File: clean_up.py (hunk 9 above)
Comment (cjc8) [reply clean_up.py:15]: This was the worst part of testing the twitter bot so kudos to you for figuring this out

File: bot.py (hunk 4 above)
Comment (maiamcc) [reply bot.py:10]: decided in meeting that this is okay, will put it in a future PR

**Pull request merged**